File: src/diffpath.cpp
#include <filesystem>
#include <iostream>
#include <set>
#include <string>
#include <vector>

namespace fs = std::filesystem;

const inline bool is_executable(const fs::path& p)
{
    std::error_code ec;
    fs::file_status status = fs::status(p, ec);
    return !ec && fs::is_regular_file(status) && ((status.permissions() & (fs::perms::owner_exec | fs::perms::group_exec | fs::perms::others_exec)) != fs::perms::none);
}

const inline std::vector<std::string> split_path(const std::string& path)
{
    std::vector<std::string> paths;
    size_t start = 0, end = 0;
    while ((end = path.find(':', start)) != std::string::npos) {
        paths.push_back(path.substr(start, end - start));
        start = end + 1;
    }
    paths.push_back(path.substr(start));
    return paths;
}

std::set<std::string> get_executables(const std::string& PATH)
{
    std::set<std::string> executables;
    for (const auto& path : split_path(PATH)) {
        try {
            for (const auto& entry : fs::directory_iterator(path)) {
                if (is_executable(entry.path())) {
                    executables.insert(entry.path().filename().string());
                }
            }
        } catch (const fs::filesystem_error& e) {
            std::cerr << "Error accessing path " << path << ": " << e.what() << std::endl;
        }
    }
    return executables;
}

void diffpath(const std::string& PATH1, const std::string& PATH2)
{
    std::set<std::string> executables1 = get_executables(PATH1);
    std::set<std::string> executables2 = get_executables(PATH2);

    // Using set_symmetric_difference to get all unique executables in sorted order
    std::vector<std::string> all_unique;
    std::set_symmetric_difference(
        executables1.begin(), executables1.end(),
        executables2.begin(), executables2.end(),
        std::back_inserter(all_unique));

    for (const auto& exec : all_unique) {
        if (executables1.contains(exec)) {
            std::cout << exec << std::endl;
        } else {
            std::cout << "\t" << exec << std::endl;
        }
    }
}

int main(int argc, char* argv[])
{
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " PATH1 PATH2" << std::endl;
        return 1;
    }

    std::string PATH1 = argv[1];
    std::string PATH2 = argv[2];

    diffpath(PATH1, PATH2);

    return 0;
}

File: src/diffpath.rs
use std::collections::BTreeSet;
use std::env;
use std::fs;
use std::os::unix::fs::PermissionsExt;
use std::path::Path;

fn is_executable(file_path: &Path) -> bool {
    if let Ok(metadata) = fs::symlink_metadata(file_path) {
        let file_type = metadata.file_type();
        if file_type.is_file() || file_type.is_symlink() {
            let permissions = metadata.permissions();
            return permissions.mode() & 0o111 != 0;
        }
    }
    false
}

fn get_executables(paths: &str) -> BTreeSet<String> {
    let mut executables = BTreeSet::new();
    for path in paths.split(':') {
        if let Ok(entries) = fs::read_dir(path) {
            for entry in entries.filter_map(Result::ok) {
                let file_path = entry.path();
                if is_executable(&file_path) {
                    if let Some(file_name) = file_path.file_name() {
                        if let Some(file_name_str) = file_name.to_str() {
                            executables.insert(file_name_str.to_string());
                        }
                    }
                }
            }
        }
    }
    executables
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: {} PATH1 PATH2", args[0]);
        std::process::exit(1);
    }

    let path1 = &args[1];
    let path2 = &args[2];

    let executables1 = get_executables(path1);
    let executables2 = get_executables(path2);

    let symmetric_difference: BTreeSet<_> = executables1
        .symmetric_difference(&executables2)
        .cloned()
        .collect();

    for exe in symmetric_difference {
        if executables1.contains(&exe) {
            println!("{}", exe);
        } else {
            println!("\t{}", exe);
        }
    }
}
